transport/src/main/java/io/netty/channel/ChannelInitializer.java

```diff
@@ -125,14 +132,25 @@ private boolean initChannel(ChannelHandlerContext ctx) throws Exception {
         return false;
     }
 
-    private void remove(ChannelHandlerContext ctx) {
+    private void remove(final ChannelHandlerContext ctx) {
         try {
             ChannelPipeline pipeline = ctx.pipeline();
             if (pipeline.context(this) != null) {
                 pipeline.remove(this);
             }
         } finally {
-            initMap.remove(ctx);
+            // The removal may happen in an async fashion if the EventExecutor we use does something funky.
+            if (ctx.isRemoved()) {
+                initMap.remove(ctx);
+            } else {
+                // Ensure we always remove from the Map in all cases to not produce a memory leak.
+                ctx.channel().closeFuture().addListener(new ChannelFutureListener() {
```

Comment:
This will retain `ctx` until the channel is closed, even if `handlerRemoved()` is called. If this is a one-off initializer (not reused) then that could greatly increase the lifetime of some objects used during handshakes and the like.

Is this seriously the only way we can clean up properly? Is the problem that the methods can be overridden? If so, then maybe we should fix that in Netty 5? Or maybe we could avoid the check-for-double-register problem by having an alternative approach to 26aa34853a8974d212e12b98e708790606bea5fa in Netty 5?

It is possible to cancel the future to clean up, but the amount of effort necessary to do something simple makes me question if this is more of systemic problem that should be addressed in Netty 5.

Reply:
@ejona86 yeah I think its the best we can do for netty 4 to ensure we not leak if the user overrides `handlerRemoved(...)`.Thats also why I check `isRemoved()` first.

Reply:
Perhaps we could check handlerRemoved() was overridden at instantiation time using reflection and take the fast path?

Reply:
@trustin we could ... that said I think it does not really worth it as `ctx.isRemoved()` should be true if the `EventExecutor` is not some funky custom implemention that is is used with `ChannelInitializer`. So I would prefer to keep things simple for now.

Reply:
Agreed.